pyo3-macros-backend/src/pyclass.rs

```diff
@@ -628,6 +759,201 @@ fn impl_enum(
     })
 }
 
+fn impl_complex_enum(
+    complex_enum: PyClassComplexEnum<'_>,
+    args: &PyClassArgs,
+    doc: PythonDoc,
+    methods_type: PyClassMethodsType,
+) -> Result<TokenStream> {
+    // Need to rig the enum PyClass options
+    let args = {
+        let mut rigged_args = args.clone();
+        // Needs to be frozen to disallow `&mut self` methods, which could break a runtime invariant
+        rigged_args.options.frozen = parse_quote!(frozen);
+        // Needs to be subclassable by the variant PyClasses
+        rigged_args.options.subclass = parse_quote!(subclass);
+        rigged_args
+    };
+
+    let krate = get_pyo3_crate(&args.options.krate);
+    let cls = complex_enum.ident;
+    let variants = complex_enum.variants;
+    let pytypeinfo = impl_pytypeinfo(cls, &args, None);
+
+    let default_slots = vec![];
+
+    let impl_builder = PyClassImplsBuilder::new(
+        cls,
+        &args,
+        methods_type,
+        complex_enum_default_methods(
+            cls,
+            variants.iter().map(|v| (v.ident, v.python_name(&args))),
+        ),
+        default_slots,
+    )
+    .doc(doc);
+
+    // Need to customize the into_py impl so that it returns the variant PyClass
+    let enum_into_py_impl = {
+        let match_arms: Vec<TokenStream> = variants
+            .iter()
+            .map(|variant| {
+                let variant_ident = variant.ident;
+                let variant_cls = gen_complex_enum_variant_class_ident(cls, variant.ident);
+                quote! {
+                    #cls::#variant_ident { .. } => {
+                        let pyclass_init = _pyo3::PyClassInitializer::from(self).add_subclass(#variant_cls);
+                        let variant_value = _pyo3::Py::new(py, pyclass_init).unwrap();
+                        _pyo3::IntoPy::into_py(variant_value, py)
+                    }
+                }
+            })
+            .collect();
+
+        quote! {
+            impl _pyo3::IntoPy<_pyo3::PyObject> for #cls {
+                fn into_py(self, py: _pyo3::Python) -> _pyo3::PyObject {
```

Comment:
So we will still have the papercut that `Py::new` doesn't match `IntoPy` for these enums. I think we should document in this PR as a known footgun and we can solve in a follow-up issue.

Reply:
Could you please explain this in more detail?
I think there should be no `Py::new` for the enum PyClass itself, but I probably misunderstand what you're saying...

Reply:
e.g. `Py::new(py, ComplexEnum::Int { i: 42 })` I think will already compile but will not produce an instance of the variant subtype, it will just create an instance of the main `ComplexEnum` base type. This is a less-used feature, so IMO I think we will get away with documenting this as slightly wrong for the first version of this feature and we can work out how to adjust later.